ext/sockets/sockets.c

```diff
@@ -2456,6 +2512,246 @@ PHP_FUNCTION(socket_export_stream)
 }
 /* }}} */
 
+/* {{{ proto resource addrinfo socket_addrinfo_lookup(string hostname[, mixed service, array hints])
+   Gets array with contents of getaddrinfo about the given hostname. */
+PHP_FUNCTION(socket_addrinfo_lookup)
+{
+	zend_string *hostname, *key;
+	zval *hint, *service, *zhints = NULL;
+
+	struct addrinfo hints, *result, *rp, *res;
+
+	memset(&hints, 0, sizeof(hints));
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "S|za", &hostname, &service, &zhints) == FAILURE) {
+		RETURN_NULL();
+	}
+
+	if (zhints) {
+		ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL_P(zhints), key, hint) {
+			if (key) {
+				if (strcmp(ZSTR_VAL(key), "ai_flags") == 0) {
+					hints.ai_flags = Z_LVAL_P(hint);
+				} else if (strcmp(ZSTR_VAL(key), "ai_socktype") == 0) {
+					hints.ai_socktype = Z_LVAL_P(hint);
+				} else if (strcmp(ZSTR_VAL(key), "ai_protocol") == 0) {
+					hints.ai_protocol = Z_LVAL_P(hint);
+				} else if (strcmp(ZSTR_VAL(key), "ai_family") == 0) {
+					hints.ai_family = Z_LVAL_P(hint);
+				}
+			}
+		} ZEND_HASH_FOREACH_END();
+	}
+
+	convert_to_string(service);
+
+	if (getaddrinfo(ZSTR_VAL(hostname), Z_STRVAL_P(service), &hints, &result) != 0) {
+		RETURN_FALSE;
+	}
+
+	array_init(return_value);
+
+	for (rp = result; rp != NULL; rp = rp->ai_next) {
+		if (rp->ai_family != AF_UNSPEC) {
+			res = emalloc(sizeof(struct addrinfo));
+			memcpy(res, rp, sizeof(struct addrinfo));
+
+			res->ai_addr = emalloc(rp->ai_addrlen);
+			memcpy(res->ai_addr, rp->ai_addr, rp->ai_addrlen);
+
+			if (rp->ai_canonname != NULL) {
+				res->ai_canonname = estrdup(rp->ai_canonname);
+			}
+
+			add_next_index_resource(return_value, zend_register_resource(res, le_addrinfo));
+		}
+	}
+
+	freeaddrinfo(result);
+}
+/* }}} */
+
+/* {{{ proto resource socket_addrinfo_bind(resource addrinfo)
+   Creates and binds to a socket from a given addrinfo resource */
+PHP_FUNCTION(socket_addrinfo_bind)
+{
+	zval			*arg1;
+	int				retval;
+	struct addrinfo	*ai;
+	php_socket		*php_sock = php_create_socket();
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "r", &arg1) == FAILURE) {
+		efree(php_sock);
```

Comment:
This case should not create and immediately destroy a socket -- better move the initialization down.